src/passes/Precompute.cpp

```diff
@@ -743,131 +745,169 @@ struct Precompute
-        // the fallthrough value's type is very different than the actually
-        // returned value's type. To handle that, if we precomputed a value and
-        // if it has the wrong type then precompute it again without looking
-        // through to the fallthrough.
-        if (values.isConcrete() &&
-            !Type::isSubType(values.getType(), set->value->type)) {
-          values = precomputeValue(set->value);
-        }
+
+    // The work list, which will contain sets and gets that have just been
+    // found to have a constant value. As we only add them to the list when they
+    // are found to be constant, each can only be added once, and a simple
+    // vector is enough here (which we can make a small vector to avoid any
+    // allocation in small-enough functions).
+    SmallVector<Expression*, 10> work;
+
+    // Given a set, see if it has a constant value. If so, note that on
+    // setValues and add to the work list.
+    auto checkConstantSet = [&](LocalSet* set) {
+      if (setValues.count(set)) {
+        // Already known to be constant.
+        return;
+      }
+
+      // Precompute the value. Note that this executes the code from scratch
+      // each time we reach this point, and so we need to be careful about
+      // repeating side effects if those side effects are expressed *in the
+      // value*. A case where that can happen is GC data (each struct.new
+      // creates a new, unique struct, even if the data is equal), and so
+      // PrecomputingExpressionRunner has special logic to make sure that
+      // reference identity is preserved properly.
+      //
+      // (Other side effects are fine; if an expression does a call and we
+      // somehow know the entire expression precomputes to a 42, then we can
+      // propagate that 42 along to the users, regardless of whatever the call
+      // did globally.)
+      auto values = precomputeValue(
+        Properties::getFallthrough(set->value, getPassOptions(), *getModule()));
+
+      // Fix up the value. The computation we just did was to look at the
+      // fallthrough, then precompute that; that looks through expressions
+      // that pass through the value. Normally that does not matter here,
+      // for example, (block .. (value)) returns the value unmodified.
+      // However, some things change the type, for example RefAsNonNull has
+      // a non-null type, while its input may be nullable. That does not
+      // matter either, as if we managed to precompute it then the value had
+      // the more specific (in this example, non-nullable) type. But there
+      // is a situation where this can cause an issue: RefCast. An attempt to
+      // perform a "bad" cast, say of a function to a struct, is a case where
+      // the fallthrough value's type is very different than the actually
+      // returned value's type. To handle that, if we precomputed a value and
```

Comment:
A cast from a function to a struct wouldn't validate. Would the comment still be correct if it used a cast between unrelated struct types as the example instead?

Reply:
Hmm, I wonder if perhaps this code was written before that change to the type hierarchies? But I think the point stands, e.g. null can be cast to non-null, which would trap at runtime.

(In any case, this part of the code is unchanged in this PR - see diff without whitespace if you haven't already.)

Reply:
I'll update the comment to fix the inaccuracy.

Reply:
Comment adjusted.

Reply:
Did you have any other comments here @tlively ?